### PR TITLE
Command to deduplicate contacts

### PR DESCRIPTION
**Pull request by alanhartless**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Automated tests included? | y
| Related user documentation PR URL | todo
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR adds a command `php app/console mautic:contacts:dedup` to find contacts with the same unique identifiers and merge them. By default, the records are merged from oldest into newest but that can be changed with the flag `--newest-into-oldest`. 

Regardless of which record is merged into which, the latest modified record's custom field data is used (except where it is empty as we will error on the side of not completely losing data). Also the latest last active date is preserved and the oldest date identified date. 

Contacts with conflicting unique identifiers when there are more than one unique identifier are not merged. 

[//]: # ( As applicable: )
#### Steps to test this PR:
1. Hack the database to update multiple contacts with the same email OR manually edit through the UI to ensure there are multiple contacts with the same unique identifiers. 
2. Run the command and they should be merged along with all of their behavior data. 

#### List deprecations along with the new alternative:
1. LeadModel::mergeLeads is deprecated. ContactMerger should be used instead

### CODE REVIEW

**Comment by dongilbert** [comment]
+1 works well. Tentative approval pending UT's

**Comment by javjim-mautic** [comment]
works properly for me

**Comment by luizeof** [comment]
@alanhartless .... 2.12.0 fix duplicating contacts on import .... this pull will be released on 2.12.0 too?

Thanks ;)

**Comment by vesper8** [comment]
I need to use this, I have a bunch of duplicate contacts and am looking for a safe way to merge them

2.12 came out already but this didn't make it in? How can I use this now? I must create a fork or is there another way?

File: app/bundles/LeadBundle/Deduplicate/ContactDeduper.php [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+{
+    /**
+     * @var FieldModel
+     */
+    private $fieldModel;
+
+    /**
+     * @var ContactMerger
+     */
+    private $merger;
```
[the remark this answers is not in the document]
Comment (alanhartless) [reply app/bundles/LeadBundle/Deduplicate/ContactDeduper.php]: No reason why not. I'll change it.

File: app/bundles/LeadBundle/Deduplicate/ContactDeduper.php [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+    /**
+     * @var LeadRepository
+     */
+    private $repository;
```
[the remark this answers is not in the document]
Comment (alanhartless) [reply app/bundles/LeadBundle/Deduplicate/ContactDeduper.php]: No reason why not. I'll change it.

File: app/bundles/LeadBundle/Exception/ValueNotMergeable.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * @author      Mautic, Inc.
+ *
+ * @link        https://mautic.org
+ *
+ * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
+ */
+
+namespace Mautic\LeadBundle\Exception;
+
+class ValueNotMergeable extends \Exception
```
[the remark this answers is not in the document]
Comment (alanhartless) [reply app/bundles/LeadBundle/Exception/ValueNotMergeable.php]: 👍

File: app/bundles/LeadBundle/Exception/ValueNotMergeable.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+{
+    /**
+     * ValueNotMergeable constructor.
+     *
+     * @param mixed $newerValue
+     * @param mixed $olderValue
+     */
+    public function __construct($newerValue, $olderValue)
+    {
+        parent::__construct(var_export($newerValue, true).' / '.var_export($olderValue, true));
```
[the remark this answers is not in the document]
Comment (alanhartless) [reply app/bundles/LeadBundle/Exception/ValueNotMergeable.php]: Just to add in the message what the value actually was. $newerValue can be anything so var_export converts it to a string for the exception message.

File: app/bundles/LeadBundle/Model/LeadModel.php [hunk 5]
```diff
@@ -2784,4 +2672,118 @@ public function setSystemCurrentLead(Lead $lead = null)
...
+     *
+     * @deprecated 2.13.0; to be removed in 3.0. Use \Mautic\LeadBundle\Model\MergeModel instead
+     *
+     * @param Lead $lead
+     * @param Lead $lead2
+     * @param bool $autoMode If true, the newest lead will be merged into the oldes then deleted; otherwise, $lead will be merged into $lead2 then deleted
+     *
+     * @return Lead
+     */
+    public function mergeLeads(Lead $lead, Lead $lead2, $autoMode = true)
```
[the remark this answers is not in the document]
Comment (alanhartless) [reply app/bundles/LeadBundle/Model/LeadModel.php]: It was relocated from elsewhere in the class. But there is a circular dependency if I use the new service here. Will have to create some kind of legacy class just to get around that.

[the remark this answers is not in the document]
Comment (alanhartless) [reply app/bundles/LeadBundle/Model/LeadModel.php]: I had to address this by passing the container to a legacy class to keep from having the circular dependency. Will likely need to leverage this class for some other methods in the LeadModel to remove more deprecated code/circular dependencies.

File: app/bundles/LeadBundle/Deduplicate/ContactMerger.php [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+     */
+    protected $dispatcher;
+
+    /**
+     * MergeModel constructor.
+     *
+     * @param LeadModel             $leadModel
+     * @param MergeRecordRepository $repo
+     * @param LoggerInterface       $logger
+     */
```
Comment (escopecz) [inline app/bundles/LeadBundle/Deduplicate/ContactMerger.php]: Dispatcher param is missing in the docblock

File: app/bundles/LeadBundle/Command/DeduplicateCommand.php [hunk 7]
```diff
@@ -0,0 +1,87 @@
+<?php
+
+/*
+ * @copyright   2017 Mautic Contributors. All rights reserved
+ * @author      Mautic, Inc.
```
[the remark this answers is not in the document]
Comment (alanhartless) [reply app/bundles/LeadBundle/Command/DeduplicateCommand.php]: Well technically it is Mautic, Inc contributing this code. So I think this is fine.

[the remark this answers is not in the document]
Comment (escopecz) [reply app/bundles/LeadBundle/Command/DeduplicateCommand.php]: Do we want every company/individual who creates/modifies the code to add itself as the author? It says "Mautic" on all files I randomly checked. Maybe @dbhurley could way in.

File: app/bundles/LeadBundle/Deduplicate/ContactDeduper.php [hunk 8]
```diff
@@ -0,0 +1,180 @@
+<?php
+
+/*
+ * @copyright   2017 Mautic Contributors. All rights reserved
```
Comment (escopecz) [inline app/bundles/LeadBundle/Deduplicate/ContactDeduper.php]: Sorry to be PITA, but could you check all the file annotations?

Comment (alanhartless) [reply app/bundles/LeadBundle/Deduplicate/ContactDeduper.php]: I think I technically wrote these files in 2017 :-)

**Pull request merged**